tests/views/test_collection_exercise.py

```diff
@@ -1375,117 +682,46 @@ def test_upload_sample(self, mock_request, mock_details):
+    @patch("response_operations_ui.views.collection_exercise.flash")
+    def test_no_upload_sample_when_no_file(self, mock_request, flash):
         sign_in_with_permission(self, mock_request, user_permission_surveys_edit_json)
         data = {"load-sample": ""}
-        with open(
-            f"{project_root}/test_data/collection_exercise/formatted_collection_exercise_details_no_sample.json"
-        ) as collection_exercise:
-            mock_details.return_value = json.load(collection_exercise)
-        mock_request.get(url_get_survey_by_short_name, status_code=200, json=self.survey_data)
-        mock_request.get(url_ces_by_survey, status_code=200, json=exercise_data)
-        response = self.client.post(
-            f"/surveys/{short_name}/{period}/upload-sample-file", data=data, follow_redirects=True
-        )
 
-        self.assertEqual(response.status_code, 200)
-        self.assertNotIn("Sample loaded successfully".encode(), response.data)
-        self.assertIn("Sample summary".encode(), response.data)
-        self.assertIn("No file uploaded".encode(), response.data)
+        self.client.post(f"/surveys/{short_name}/{period}/upload-sample-file", data=data)
+        flash.assert_called_once_with("No file uploaded", "error")
 
+    @mock_decorator(CE, SEFT_SURVEY)
     @requests_mock.mock()
-    @patch("response_operations_ui.views.collection_exercise.build_collection_exercise_details")
-    def test_upload_sample_csv_too_few_columns(self, mock_request, mock_details):
+    @patch("response_operations_ui.controllers.sample_controllers.upload_sample")
+    @patch("response_operations_ui.views.collection_exercise.flash")
+    def test_upload_sample_csv_too_few_columns(self, mock_request, flash, upload_sample):
         sign_in_with_permission(self, mock_request, user_permission_surveys_edit_json)
         post_data = {"sampleFile": (BytesIO(b"data"), "test.csv")}
+        exception = ApiError(fake_response(url="", status_code=400, text="Too few columns in CSV file"))
+        upload_sample.side_effect = exception
 
-        with open(
-            f"{project_root}/test_data/collection_exercise/formatted_collection_exercise_details_no_sample.json"
-        ) as collection_exercise:
-            mock_details.return_value = json.load(collection_exercise)
-        mock_request.get(url_ces_by_survey, json=self.collection_exercises)
-        mock_request.get(url_get_survey_by_short_name, json=self.survey_data)
-        mock_request.get(url_ces_by_survey, json=exercise_data)
-        mock_request.post(url_sample_service_upload, status_code=400, text="Too few columns in CSV file")
+        self.client.post(f"/surveys/{short_name}/{period}/upload-sample-file", data=post_data)
 
-        response = self.client.post(
-            f"/surveys/{short_name}/{period}/upload-sample-file", data=post_data, follow_redirects=True
-        )
+        flash.assert_called_once_with("Too few columns in CSV file", "error")
```

Comment:
Quite often we are following url redirects to the main CE page, we should avoid that if possible, it just slows everything down if that page has already been tested, asserted and adds nothing more. We should just assert on the action, in this instance a flash message